### PR TITLE
Normalize PyPI package names

### PR DESCRIPTION
**Pull request by andrewpollock**
Fixes: #508

### CODE REVIEW

File: osv/models.py [hunk 1]
```diff
@@ -332,18 +332,25 @@ def _tokenize(self, value):
     value_lower = value.lower()
     return re.split(r'\W+', value_lower) + [value_lower]
 
+  def _normalize(self, name):
+    """Normalize package name per https://peps.python.org/pep-0503/#normalized-names"""
+
```
Comment (oliverchang) [inline osv/models.py:337]: nit: remove this newline.

File: osv/models.py [hunk 2]
```diff
@@ -332,18 +332,25 @@ def _tokenize(self, value):
...
+    return re.sub(r"[-_.]+", "-", name).lower()
```
Comment (oliverchang) [inline osv/models.py:338]: nit: use single quotes rather than double quotes.

File: osv/models.py [hunk 3]
```diff
@@ -332,18 +332,25 @@ def _tokenize(self, value):
     value_lower = value.lower()
     return re.split(r'\W+', value_lower) + [value_lower]
 
+  def _normalize(self, name):
```
Comment (oliverchang) [inline osv/models.py:335]: nit: s/_normalize/_normalize_pypi/ to be clearer. (but see below for another comment which makes this no longer relevant).

File: osv/models.py [hunk 4]
```diff
@@ -332,18 +332,25 @@ def _tokenize(self, value):
...
+    for pkg in self.affected_packages:
       # Set PURL if it wasn't provided.
-      if not affected_package.package.purl:
-        affected_package.package.purl = purl_helpers.package_to_purl(
+      if not pkg.package.purl:
+        pkg.package.purl = purl_helpers.package_to_purl(
             ecosystems.normalize(affected_package.package.ecosystem),
-            affected_package.package.name)
+            pkg.package.name)
+      if pkg.package.ecosystem and pkg.package.ecosystem == 'PyPI':
```
Comment (oliverchang) [inline osv/models.py:352]: nit: no need for the first part of the if, as if it's not set the second comparison will fail anyway. 

That said, I would suggest refactoring this into its own function like so:

```python
def _normalize_package(package):
  """docstring here"""
  if package.ecosystem == 'PyPI':
     # per link...
    return re.sub(r"[-_.]+", "-", package.name).lower()

#  ... in the pre_put_hook 
for pkg in self.affected_packages:
  pkg.package = _normalize_package(pkg.package)
```

File: osv/models.py [hunk 5]
```diff
@@ -332,18 +332,25 @@ def _tokenize(self, value):
...
     search_indices = set()
 
     search_indices.update(self._tokenize(self.id()))
 
-    for affected_package in self.affected_packages:
+    for pkg in self.affected_packages:
       # Set PURL if it wasn't provided.
-      if not affected_package.package.purl:
-        affected_package.package.purl = purl_helpers.package_to_purl(
+      if not pkg.package.purl:
```
Comment (oliverchang) [inline osv/models.py:348]: This should go after your normalization, as otherwise the PURL won't get the correct normalized value.

**Review by oliverchang — COMMENTED** [review]
nice! some comments.

Please also update the test here: https://github.com/google/osv.dev/blob/a4b682a32575cc3314a5ef83c8e91b70c60f7b77/docker/worker/worker_test.py#L891

to make the test data have an un-normalized name.

File: osv/models.py (hunk 1 above)
Comment (andrewpollock) [reply osv/models.py:340]: Addressed in 47438b76ae7c3b24ee97d85478bc01325d888bb6

File: osv/models.py (hunk 2 above)
Comment (andrewpollock) [reply osv/models.py]: Addressed in 47438b76ae7c3b24ee97d85478bc01325d888bb6

File: osv/models.py (hunk 3 above)
Comment (andrewpollock) [reply osv/models.py]: Ack. Addressed in 47438b76ae7c3b24ee97d85478bc01325d888bb6

File: osv/models.py (hunk 4 above)
Comment (andrewpollock) [reply osv/models.py]: Addressed in 47438b76ae7c3b24ee97d85478bc01325d888bb6

File: osv/models.py (hunk 5 above)
Comment (andrewpollock) [reply osv/models.py:352]: Ah, good catch, thanks. Addressed in 47438b76ae7c3b24ee97d85478bc01325d888bb6

**Review by oliverchang — COMMENTED** [review]
Nice! Looking good. We just need that test now :)

**Comment by andrewpollock** [comment]
By George, I do believe I have tests. PTAL

**Comment by oliverchang** [comment]
Very nice! I think you just need to format your code per https://github.com/google/osv.dev/runs/7782343232?check_suite_focus=true.

(Ignore the lint errors -- this wasn't blocking our check and we need to fix them per https://github.com/google/osv.dev/issues/521)

**Comment by andrewpollock** [comment]
My bad, I knew I'd introduced changes that needed that and forgot to run it. PTAL.

**Review by oliverchang — APPROVED** [review]
very nice!! LGTM

**Pull request merged**